src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -486,20 +456,14 @@ else if (null != mHeader) {
      * sequence, or if the mate reference name is not found in the sequence index, NO_ALIGNMENT_REFERENCE_INDEX (-1)
      * is returned.
      *
-     * @throws IllegalStateException if the mate reference index cannot be resolved because the SAMFileHeader for the
-     * record is null.
+     * @throws IllegalStateException if the mate reference index must be resolved but cannot be because the
+     * SAMFileHeader for the record is null.
      */
     public Integer getMateReferenceIndex() {
-        if (null == mMateReferenceIndex) {
-            // try to resolve the reference index
-            if (NO_ALIGNMENT_REFERENCE_NAME.equals(mMateReferenceName)) {
-                mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            }
-            else if (null != mHeader) {
-                mMateReferenceIndex = mHeader.getSequenceIndex(mMateReferenceName);
-            }
-            else {
-                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the mate reference index");
+        if (null == mMateReferenceIndex) { // try to resolve the mate reference index
+            mMateReferenceIndex = resolveIndexFromName(mMateReferenceName, mHeader, false);
+            if (null == mMateReferenceIndex) {
+                return NO_ALIGNMENT_REFERENCE_INDEX;
```

Comment:
Same comment as for `getReferenceIndex()` -- This new code leaves mMateReferenceIndex set to null in the case where the lookup failed, whereas the previous version would leave it set to `NO_ALIGNMENT_REFERENCE_INDEX`. Do we care?